### ISSUE TITLE
Explore Kotlin 1.1 features

### ISSUE THREAD

**Issue by czyzby** [czyzby/ktx#40]
- ~~Builder scoping DSL could greatly improve Scene2D and VisUI building APIs.~~ #44 #46 
- ~~Builder scoping DSL for `Skin` building methods.~~ #45 
- [x] Type aliases for LibGDX collections could help avoiding collisions with Kotlin classes (`Array`, for example), as well as providing class names consistent with the factory methods.
- ~~Coroutines could be potentially used for asset loading and delegating tasks to main loop (similarly to official Swing utilities library).~~ #41 #47 
- [x] Some properties might be inlined (especially extensions).

**Comment by sreich** [comment]
I've been especially curious on coroutines. Right now I'm just trying to use them in my game to multithread the world gen.

But I'm really interested (and confused) in how they can be used well throughout a game(where it's single threaded), aside from the obvious asset stuff.

We can write our own dispatcher, for libgdx which can run the tasks on the UI thread using libgdx mechanisms for that.

But mostly I'm just trying to think of situations where they make sense to apply, as a game. They're rather new in general, few other languages support them and game industry seems to be slower at adopting things, so there is not much material on the subject. Plus, none in mainstream are as flexible as kotlin's.. So it's a bit new territory.

**Comment by czyzby** [comment]
> But I'm really interested (and confused) in how they can be used well throughout a game(where it's single threaded), aside from the obvious asset stuff.

Single-threaded applications will not benefit from coroutines. Most applications won't, to be honest. Assets can already be loaded asynchronously, and you don't really want to pause your main rendering logic in order to load some asset on a separate thread.

However, I do see a use case where you could schedule handling of some networking packets during the main rendering loop without blocking the whole networking thread. Currently this can be achieved with callbacks, which are obviously more tedious to use.

```Kotlin
// Coroutines:
fun handleSomePacket(packet: ImportantServerPacket) {
  val response = gui.handlePacket(packet).await()
  server.send(response)
}

// ...vs callbacks:
fun handleSomePacket(packet: ImportantServerPacket) {
  Gdx.app.postRunnable {
    val response = gui.handlePacket(packet)
    networkingExecutor.run {
      server.send(response)
    }
  }
}
// Not as bad as Java <8, but still.
```

It could be also useful for other events that happen outside of rendering thread scope. For example, you could write a custom asset loader for `AssetManager` that sends a notification to the rendering thread to display a toast each time an important asset is loaded. It would be more convenient that checking if the asset is loaded on each render.

To be honest, I haven't fully explored coroutines API yet, so it seems like a good opportunity. If the utilities turn out somewhat useful, I'll put them in a separate module.

**Comment by sreich** [comment]
>Single-threaded applications will not benefit from coroutines. 

that's not the conclusion i came to so far. looks like they can be used to simplify/mask synchronous callback logic or polling. there are synchronous and asynchronous coroutines. you can do whatever you want with them and you can decide however you want to schedule them (either on n threads or on the gl thread, when...how to handle it, it's up to you etc..)

http://www.ilikebigbits.com/blog/2016/3/20/coroutines-for-games

unity's example shows you using them to do a fade on something.

>ssets can already be loaded asynchronously, and you don't really want to pause your main rendering logic in order to load some asset on a separate thread.

oh no, kotlin is extremely flexible with coroutines. the features built into the language are mostly all low level and in the stdlib (there is 1 keyword i think is all, the rest are regular functions), then there's the kotlinx.coroutines which is just a regular library built on top of those primitives meant to make it easier to interact with and provide more common functionality, with some bindings to things like javafx and the like.

**Comment by czyzby** [comment]
I'm not sure if you can easily suspend rendering thread and if that makes any sense. Most platforms just run an endless loop on the main thread - that might be problematic. After all, you cannot just suspend and skip a rendering call, only to come back to if after some action on other thread as finished. Game with actual logic represented by coroutines (like in your link) would have be designed to leverage the coroutines API from day one. It simply might be out of scope of KTX.

Anyway, I'll have to look into it and see what can be done, that's all I can say for now.

**Comment by sreich** [comment]
>After all, you cannot just suspend and skip a rendering call, only to come back to if after some action on other thread as finished

yeah, you'd only be suspending functions and picking them up later that happen to be on the render thread but aren't rendery things themselves (the gl state would get screwed with, as you note). but rendering is just a dumb operation anyways, data modification is where it'd be interesting (in case of the fade example, that would be a resumption of a suspended function that sets the fade level of an actor)

>It simply might be out of scope of KTX.
yeah i agree, i was just kinda talking aloud. the most ktx could do is help give you dispatchers/some syntax to facilitate some usages of them with libgdx.

**Comment by czyzby** [comment]
I'm afraid `@DslMarker` annotation will not help current `Skin` and `Scene2D` builders, since they extend existing APIs of LibGDX classes to provide builder-like DSL. DSL marker annotations either do not work with extension methods or need annotations on classes that we do not control (original LibGDX API).

**Comment by kotcrab** [comment]
I'm trying to use it now, seems to be possible since we are extending most of (all? I don't really remember) libgdx widgets. 

Check this modified [`widgets.kt`](https://gist.github.com/kotcrab/6615a7480c8482d3872455f78af49563) from `vis` module. I only declared `@VisTagMarker` and added it to `KTabbedPane`, `KTab`. Now try this example:

```kotlin
table {
    tabbedPane {
        tab("Tab1") {
            label("Inside tab 1")
            tab("Tab") { // illegal now
            }
        }

        tab("Tab2") {
        }
    }
}
```
It's weird how this annotation works, still trying to wrap my head around it.

**Comment by czyzby** [comment]
Huh. I tried to use it on `KTable` interface from `scene2d` and `skin` method from `style` module and both didn't work as expected. I think the difference between `vis` and `scene2d` modules is the fact that all factory methods in `scene2d` are inlined extension methods as opposed to actual methods of the builder classes in `vis`. I'll give it another try later.

Anyway, I'm currently implementing `ktx-async` with coroutines support based on the Swing example from JetBrains. If it works, this is a game changer for asynchronous APIs in LibGDX.

**Comment by czyzby** [comment]
Basic coroutines support will be introduced in `ktx-async` module. Asset management improvements will be implemented in #41.

**Comment by czyzby** [comment]
Turns out you can use `@DslMarker` even if you do not control the class by applying the annotation on parameter types of functions. This means we can update both `ktx-style` and `ktx-scene2d` without major refactoring and - hopefully - breaking API changes.

This issue turned out to be rather huge, so it was separated into multiple tasks. I'm closing this for now.
